File: Go/pkg/component/prototype.go
package component

/*
	Package 'component' provides definitions for various objects meeting the definition of a protocol component.
*/

import (
	"donp/pkg/element"
	"errors"
	"log/slog"
)

/*
	Objects of type Prototype represent the structure of a particular message type within a protocol.
	Required fields: name, transmit, receive.
	Optional fields: desc.
*/

type Prototype struct {
	Name       string             `json:"name"`
	Desc       string             `json:"desc"`
	TxSegments []*element.Segment `json:"transmit"`
	RxSegments []*element.Segment `json:"receive"`
}

func NewPrototype(dataMap map[string]interface{}) (*Prototype, error) {
	var prototype Prototype
	if name, ok := dataMap["name"].(string); ok {
		prototype.Name = name
	} else {
		return nil, errors.New("Prototype: name field is required")
	}
	if desc, ok := dataMap["desc"].(string); ok {
		prototype.Desc = desc
	}
	var err error
	// Build transmit prototype segments
	prototype.TxSegments, err = prototype.initSegments("transmit", dataMap)
	if err != nil {
		return nil, err
	}
	// Build receive prototype segments
	prototype.RxSegments, err = prototype.initSegments("receive", dataMap)
	if err != nil {
		return nil, err
	}
	return &prototype, nil
}

func (p *Prototype) initSegments(name string, dataMap map[string]interface{}) ([]*element.Segment, error) {
	segments := []*element.Segment{}
	if segs, ok := dataMap[name].([]interface{}); ok {
		for _, seg := range segs {
			if segMap, ok := seg.(map[string]interface{}); ok {
				segment, err := element.NewSegment(segMap)
				if err != nil {
					slog.Error("Failed to create " + name + " segment: " + err.Error())
					continue
				}
				segments = append(segments, segment)
			}
		}
	} else {
		return nil, errors.New("Prototype: " + name + " field is required")
	}
	return segments, nil
}

// Log outputs prototype information
func (p *Prototype) Log() {
	slog.Debug("Prototype: " + p.Name + " (" + p.Desc + ")")
}

// GetSegments retrieves the segments based on the specified direction.
func (p *Prototype) GetSegments(direction Direction) []*element.Segment {
	switch direction {
	case DirectionTX:
		return p.TxSegments
	case DirectionRX:
		return p.RxSegments
	}
	slog.Error("Invalid direction")
	return nil
}

File: Python/pkg/component/prototype.py
"""
    Package 'component' provides definitions for various objects meeting the definition of a protocol component.
"""

from pkg.element.segment import Segment
from constants import Direction
import json

"""
    Objects of class Prototype represent the structure of a particular message type within a protocol.
    Required fields include name, transmit segments, and receive segments.
    Optional fields include desc.

"""

class Prototype:

    def __init__(self, protocol: json, logger):
        # Setup logger
        self.logger = logger
        # Initialize prototype fields
        self.tx = []
        self.rx = []
        # Ensure all required fields are present
        if 'name' in protocol:
            if 'transmit' in protocol:
                if 'receive' in protocol:
                    # Initialize optional prototype fields       
                    desc = ""
                    # Populate prototype fields if they exist
                    if 'name' in protocol:
                        name = protocol['name']
                    if 'desc' in protocol:
                        desc = protocol['desc']
                    self.__initialize(name, desc, protocol['transmit'], protocol['receive'])
                else:
                    raise AttributeError("Prototype: receive field is required")
            else:
                raise AttributeError("Prototype: transmit field is required")
        else:
            raise AttributeError("Prototype: name field is required")
    

    def __initialize(self, name: str, desc: str, transmit: json, receive: json):
        self.name = name
        self.desc = desc
        for seg in transmit:
            segment = Segment(seg, self.logger)
            self.tx.append(segment)
        for seg in receive:
            segment = Segment(seg, self.logger)
            self.rx.append(segment)

    def log(self):
        """
        Logs prototype information.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.logger.debug(f"Prototype: " + self.name + " (" + self.desc + ")")

    def get_segments(self, direction: Direction):
        """
        Retrieves segments based on the specified direction.

        Args:
            direction (Direction): The direction of the segments to retrieve.
        Returns:
            list: A list of segments corresponding to the direction.
        Raises:
            ValueError: If the direction is invalid.
        """
        if direction == Direction.TX:
            return self.tx
        elif direction == Direction.RX:
            return self.rx
        else:
            raise ValueError(f"Invalid direction specified {direction}")
   